These error occur here:
1. test_reproduce_turn_on(hass)
https://github.com/home-assistant/home-assistant/blob/8ba4ee1012fcdbfe7b238ef1ac4de34e09adf641/tests/helpers/test_state.py#L132
```python
>       assert ["light.test"] == last_call.data.get("entity_id")
E       AssertionError: assert ['light.test'] == 'light.test'
```

2. test_reproduce_turn_off(hass)
https://github.com/home-assistant/home-assistant/blob/8ba4ee1012fcdbfe7b238ef1ac4de34e09adf641/tests/helpers/test_state.py#L149
```python
>       assert ["light.test"] == last_call.data.get("entity_id")
E       AssertionError: assert ['light.test'] == 'light.test'
```

3. test_reproduce_complex_data(hass)
https://github.com/home-assistant/home-assistant/blob/8ba4ee1012fcdbfe7b238ef1ac4de34e09adf641/tests/helpers/test_state.py#L170
```python
>       assert complex_data == last_call.data.get("complex")
E       AssertionError: assert ['hello', {'11': '22'}] == None
```

Unfortunately I've no idea why these tests fail.

--- a/homeassistant/components/light/reproduce_state.py
+++ b/homeassistant/components/light/reproduce_state.py
@@ -33,6 +33,15 @@ from . import (
 _LOGGER = logging.getLogger(__name__)
 
 VALID_STATES = {STATE_ON, STATE_OFF}
+
+ATTR_GROUP = [
+    ATTR_BRIGHTNESS,
+    ATTR_EFFECT,
+    ATTR_FLASH,
+    ATTR_WHITE_VALUE,
+    ATTR_TRANSITION,
+]
+
 COLOR_GROUP = [
     ATTR_COLOR_NAME,
     ATTR_COLOR_TEMP,
@@ -83,18 +92,13 @@ async def _async_reproduce_state(
 
     if state.state == STATE_ON:
         service = SERVICE_TURN_ON
-        if ATTR_BRIGHTNESS in state.attributes:
-            service_data[ATTR_BRIGHTNESS] = state.attributes[ATTR_BRIGHTNESS]
-        if ATTR_EFFECT in state.attributes:
-            service_data[ATTR_EFFECT] = state.attributes[ATTR_EFFECT]
-        if ATTR_FLASH in state.attributes:
-            service_data[ATTR_FLASH] = state.attributes[ATTR_FLASH]
-        if ATTR_TRANSITION in state.attributes:
-            service_data[ATTR_TRANSITION] = state.attributes[ATTR_TRANSITION]
-        if ATTR_WHITE_VALUE in state.attributes:
-            service_data[ATTR_WHITE_VALUE] = state.attributes[ATTR_WHITE_VALUE]
+        for attr in ATTR_GROUP:
+            # All attributes that are not colors
+            if attr in state.attributes:
+                service_data[attr] = state.attributes[attr]
 
         for color_attr in COLOR_GROUP:
+            # Choose the first color that is specified
             if color_attr in state.attributes:
                 service_data[color_attr] = state.attributes[color_attr]
                 break

--- a/tests/components/light/test_reproduce_state.py
+++ b/tests/components/light/test_reproduce_state.py
@@ -90,56 +90,67 @@ async def test_reproducing_states(hass, caplog):
 
     assert len(turn_on_calls) == 12
 
-    for i in range(0, 12):
-        assert turn_on_calls[i].domain == "light"
+    expected_calls = []
 
     expected_off = VALID_BRIGHTNESS
     expected_off["entity_id"] = "light.entity_off"
-    assert turn_on_calls[0].data == expected_off
+    expected_calls.append(expected_off)
 
     expected_bright = VALID_WHITE_VALUE
     expected_bright["entity_id"] = "light.entity_bright"
-    assert turn_on_calls[1].data == expected_bright
+    expected_calls.append(expected_bright)
 
     expected_white = VALID_FLASH
     expected_white["entity_id"] = "light.entity_white"
-    assert turn_on_calls[2].data == expected_white
+    expected_calls.append(expected_white)
 
     expected_flash = VALID_EFFECT
     expected_flash["entity_id"] = "light.entity_flash"
-    assert turn_on_calls[3].data == expected_flash
+    expected_calls.append(expected_flash)
 
     expected_effect = VALID_TRANSITION
     expected_effect["entity_id"] = "light.entity_effect"
-    assert turn_on_calls[4].data == expected_effect
+    expected_calls.append(expected_effect)
 
     expected_trans = VALID_COLOR_NAME
     expected_trans["entity_id"] = "light.entity_trans"
-    assert turn_on_calls[5].data == expected_trans
+    expected_calls.append(expected_trans)
 
     expected_name = VALID_COLOR_TEMP
     expected_name["entity_id"] = "light.entity_name"
-    assert turn_on_calls[6].data == expected_name
+    expected_calls.append(expected_name)
 
     expected_temp = VALID_HS_COLOR
     expected_temp["entity_id"] = "light.entity_temp"
-    assert turn_on_calls[7].data == expected_temp
+    expected_calls.append(expected_temp)
 
     expected_hs = VALID_KELVIN
     expected_hs["entity_id"] = "light.entity_hs"
-    assert turn_on_calls[8].data == expected_hs
+    expected_calls.append(expected_hs)
 
     expected_kelvin = VALID_PROFILE
     expected_kelvin["entity_id"] = "light.entity_kelvin"
-    assert turn_on_calls[9].data == expected_kelvin
+    expected_calls.append(expected_kelvin)
 
     expected_profile = VALID_RGB_COLOR
     expected_profile["entity_id"] = "light.entity_profile"
-    assert turn_on_calls[10].data == expected_profile
+    expected_calls.append(expected_profile)
 
     expected_rgb = VALID_XY_COLOR
     expected_rgb["entity_id"] = "light.entity_rgb"
-    assert turn_on_calls[11].data == expected_rgb
+    expected_calls.append(expected_rgb)
+
+    for call in turn_on_calls:
+        assert call.domain == "light"
+        found = False
+        for expected in expected_calls:
+            if call.data["entity_id"] == expected["entity_id"]:
+                # We found the matching entry
+                assert call.data == expected
+                found = True
+                break
+        # No entry found
+        assert found
 
     assert len(turn_off_calls) == 1
     assert turn_off_calls[0].domain == "light"

--- a/tests/helpers/test_state.py
+++ b/tests/helpers/test_state.py
@@ -129,7 +129,7 @@ async def test_reproduce_turn_on(hass):
     last_call = calls[-1]
     assert last_call.domain == "light"
     assert SERVICE_TURN_ON == last_call.service
-    assert ["light.test"] == last_call.data.get("entity_id")
+    assert "light.test" == last_call.data.get("entity_id")
 
 
 async def test_reproduce_turn_off(hass):
@@ -146,7 +146,7 @@ async def test_reproduce_turn_off(hass):
     last_call = calls[-1]
     assert last_call.domain == "light"
     assert SERVICE_TURN_OFF == last_call.service
-    assert ["light.test"] == last_call.data.get("entity_id")
+    assert "light.test" == last_call.data.get("entity_id")
 
 
 async def test_reproduce_complex_data(hass):
@@ -155,10 +155,10 @@ async def test_reproduce_complex_data(hass):
 
     hass.states.async_set("light.test", "off")
 
-    complex_data = ["hello", {"11": "22"}]
+    complex_data = [255, 100, 100]
 
     await state.async_reproduce_state(
-        hass, ha.State("light.test", "on", {"complex": complex_data})
+        hass, ha.State("light.test", "on", {"rgb_color": complex_data})
     )
 
     await hass.async_block_till_done()
@@ -167,7 +167,7 @@ async def test_reproduce_complex_data(hass):
     last_call = calls[-1]
     assert last_call.domain == "light"
     assert SERVICE_TURN_ON == last_call.service
-    assert complex_data == last_call.data.get("complex")
+    assert complex_data == last_call.data.get("rgb_color")
 
 
 async def test_reproduce_bad_state(hass):

